From 798d80499a41a92d12ad57d169ed0db2fcac4a5e Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Thu, 21 Nov 2024 19:02:15 +0100
Subject: [PATCH] [DogeCash] `CheckAuxProofOfWork`: Validate AuxPow (#53)

Dogecoin supports merge-mining through AuxPow. This is a continuation of previous work, and now we actually allow AuxPow on blocks that have it.

(cherry picked from commit cf5f71083c18de0dd9f95e6f523e9f6907c98d52)

diff --git a/src/pow/auxpow.cpp b/src/pow/auxpow.cpp
index cd87b0efae260d5c1572e10d87eb79cd1c28e258..120d712946bf276c22c60118aff53a2a5722fb59 100644
--- a/src/pow/auxpow.cpp
+++ b/src/pow/auxpow.cpp
@@ -21,8 +21,35 @@ bool CheckAuxProofOfWork(const CBlockHeader &block,
                      block.nVersion);
     }
 
-    if (!CheckProofOfWork(block.GetPowHash(), block.nBits, params)) {
-        return error("%s: non-AUX proof of work failed", __func__);
+    // If there is no auxpow, just check the block hash.
+    if (!block.auxpow) {
+        if (VersionHasAuxPow(block.nVersion)) {
+            return error("%s: no auxpow on block %s with auxpow version %08x",
+                         __func__, block.GetHash().ToString(), block.nVersion);
+        }
+
+        if (!CheckProofOfWork(block.GetPowHash(), block.nBits, params)) {
+            return error("%s: non-AUX proof of work failed", __func__);
+        }
+
+        return true;
+    }
+
+    if (!VersionHasAuxPow(block.nVersion)) {
+        // Header encodes auxpow, but version doesn't reflect it
+        return error("%s: AuxPow on block with non-auxpow version", __func__);
+    }
+
+    util::Result<std::monostate> auxResult = block.auxpow->CheckAuxBlockHash(
+        block.GetHash(), VersionChainId(block.nVersion), params);
+    if (!auxResult) {
+        return error("%s: AuxPow validity check failed: %s", __func__,
+                     ErrorString(auxResult).original);
+    }
+
+    if (!CheckProofOfWork(BlockHash(block.auxpow->parentBlock.GetPowHash()),
+                          block.nBits, params)) {
+        return error("%s: Auxillary header proof of work failed", __func__);
     }
 
     return true;
diff --git a/src/test/CMakeLists.txt b/src/test/CMakeLists.txt
index 27ec03b8e9e4d57abe96838c32d9c06f41d61264..15832a4fe195fb6f679aeb8ab474008d09775d4d 100644
--- a/src/test/CMakeLists.txt
+++ b/src/test/CMakeLists.txt
@@ -177,6 +177,7 @@ add_boost_unit_tests_to_suite(bitcoin test_bitcoin
 		dogecoin_block_version_tests.cpp
 		dogecoin_auxpow_check_tests.cpp
 		dogecoin_auxpow_methods_tests.cpp
+		dogecoin_auxpow_block_tests.cpp
 		dogecoin_auxpow_serialize_tests.cpp
 		dstencode_tests.cpp
 		feefrac_tests.cpp
diff --git a/src/test/dogecoin_auxpow_block_tests.cpp b/src/test/dogecoin_auxpow_block_tests.cpp
new file mode 100644
index 0000000000000000000000000000000000000000..9c022bf644d744432a3e2209e01c6bbfcaf722c4
--- /dev/null
+++ b/src/test/dogecoin_auxpow_block_tests.cpp
@@ -0,0 +1,137 @@
+// Copyright (c) 2024 The Bitcoin developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <pow/auxpow.h>
+#include <pow/pow.h>
+#include <primitives/auxpow.h>
+#include <validation.h>
+
+#include <test/lcg.h>
+#include <test/util/setup_common.h>
+
+#include <boost/test/unit_test.hpp>
+
+BOOST_FIXTURE_TEST_SUITE(dogecoin_auxpow_block_tests, TestChain100Setup)
+
+BOOST_AUTO_TEST_CASE(mine_auxpow_blocks_test) {
+    ChainstateManager &chainman = *Assert(m_node.chainman);
+    const Config &config = GetConfig();
+    const Consensus::Params &consensus = config.GetChainParams().GetConsensus();
+    const BlockValidationOptions blockopts(config);
+
+    CBlock block =
+        CreateAuxPowBlock({}, CScript() << OP_1, 0x63, 0x12345678, {uint256()},
+                          {uint256(), uint256()}, chainman.ActiveChainstate());
+
+    // Block must have the auxpow version bit set if it has auxpow
+    {
+        CBlock block = CreateAuxPowBlock(
+            {}, CScript() << OP_1, 0x63, 0x12345678, {uint256()},
+            {uint256(), uint256()}, chainman.ActiveChainstate());
+        block.nVersion = VersionWithAuxPow(block.nVersion, false);
+        BlockValidationState state;
+        BOOST_CHECK(!CheckBlock(block, state, consensus, blockopts));
+        BOOST_CHECK_EQUAL(state.ToString(), "high-hash, proof of work failed");
+    }
+
+    // Block must NOT have the auxpow version bit set if it does NOT has auxpow
+    {
+        CBlock block = CreateAuxPowBlock(
+            {}, CScript() << OP_1, 0x63, 0x12345678, {uint256()},
+            {uint256(), uint256()}, chainman.ActiveChainstate());
+        block.auxpow.reset();
+        BlockValidationState state;
+        BOOST_CHECK(!CheckBlock(block, state, consensus, blockopts));
+        BOOST_CHECK_EQUAL(state.ToString(), "high-hash, proof of work failed");
+    }
+
+    // CheckAuxBlockHash failed: Parent chain ID can't be our chain ID
+    {
+        CBlock block = CreateAuxPowBlock(
+            {}, CScript() << OP_1, AUXPOW_CHAIN_ID, 0x12345678, {uint256()},
+            {uint256(), uint256()}, chainman.ActiveChainstate());
+        BlockValidationState state;
+        BOOST_CHECK(!CheckBlock(block, state, consensus, blockopts));
+        BOOST_CHECK_EQUAL(state.ToString(), "high-hash, proof of work failed");
+    }
+
+    // CheckAuxBlockHash failed: nIndex must be 0
+    {
+        CBlock block = CreateAuxPowBlock(
+            {}, CScript() << OP_1, 0x63, 0x12345678, {uint256()},
+            {uint256(), uint256()}, chainman.ActiveChainstate());
+        block.auxpow->nIndex = 1;
+        BlockValidationState state;
+        BOOST_CHECK(!CheckBlock(block, state, consensus, blockopts));
+        BOOST_CHECK_EQUAL(state.ToString(), "high-hash, proof of work failed");
+    }
+
+    // High-hash on parent block
+    {
+        CBlock block = CreateAuxPowBlock(
+            {}, CScript() << OP_1, 0x63, 0x12345678, {uint256()},
+            {uint256(), uint256()}, chainman.ActiveChainstate());
+
+        // Ensure parent block's nonce DOESNT mine the block
+        while (CheckProofOfWork(block.auxpow->parentBlock.GetPowHash(),
+                                block.nBits, consensus)) {
+            ++block.auxpow->parentBlock.nNonce;
+        }
+
+        BlockValidationState state;
+        BOOST_CHECK(!CheckBlock(block, state, consensus, blockopts));
+        BOOST_CHECK_EQUAL(state.ToString(), "high-hash, proof of work failed");
+    }
+
+    // Valid block
+    {
+        CBlock block = CreateAuxPowBlock(
+            {}, CScript() << OP_1, 0x63, 0x12345678, {uint256()},
+            {uint256(), uint256()}, chainman.ActiveChainstate());
+        BlockValidationState state;
+        BOOST_CHECK(CheckBlock(block, state, consensus, blockopts));
+    }
+}
+
+uint256 GenHash(MMIXLinearCongruentialGenerator &lcg) {
+    uint256 hash;
+    for (uint8_t &byte : hash) {
+        byte = lcg.next();
+    }
+    return hash;
+}
+
+BOOST_AUTO_TEST_CASE(mine_auxpow_blocks_rng_test) {
+    MMIXLinearCongruentialGenerator lcg;
+
+    // Randomly generate a lot of configurations and test for successes.
+    for (size_t testCase = 0; testCase < 100; testCase++) {
+        // Generate random parameters
+
+        uint32_t parentChainId = lcg.next() % 0x10000;
+        if (parentChainId == AUXPOW_CHAIN_ID) {
+            // Cannot have parent chain ID equal to our chain ID
+            parentChainId = (parentChainId + 1) % 0x10000;
+        }
+
+        const uint32_t chainMerkleHeight = lcg.next() % 31;
+        const uint32_t nMergeMineNonce = lcg.next();
+        const uint32_t coinbaseMerkleHeight = lcg.next() % 32;
+
+        std::vector<uint256> chainMerkleBranch;
+        for (size_t i = 0; i < chainMerkleHeight; ++i) {
+            chainMerkleBranch.push_back(GenHash(lcg));
+        }
+        std::vector<uint256> coinbaseMerkleBranch;
+        for (size_t i = 0; i < coinbaseMerkleHeight; ++i) {
+            coinbaseMerkleBranch.push_back(GenHash(lcg));
+        }
+
+        CreateAndProcessAuxPowBlock({}, CScript() << OP_1, parentChainId,
+                                    nMergeMineNonce, chainMerkleBranch,
+                                    coinbaseMerkleBranch);
+    }
+}
+
+BOOST_AUTO_TEST_SUITE_END()
diff --git a/src/test/dogecoin_auxpow_check_tests.cpp b/src/test/dogecoin_auxpow_check_tests.cpp
index bd60a79f077f2f6371ee4d1ab1ed19cb77116034..9fd9ded920ab87ca06c0dbafe9563c4479c0932b 100644
--- a/src/test/dogecoin_auxpow_check_tests.cpp
+++ b/src/test/dogecoin_auxpow_check_tests.cpp
@@ -2,6 +2,7 @@
 // Distributed under the MIT software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
+#include <pow/auxpow.h>
 #include <primitives/auxpow.h>
 #include <span.h>
 #include <streams.h>
@@ -105,6 +106,8 @@ BOOST_AUTO_TEST_CASE(auxpow_block700000_test) {
     Consensus::Params params = CChainParams::Main({})->GetConsensus();
     BOOST_CHECK(auxpow.CheckAuxBlockHash(
         header.GetHash(), VersionChainId(header.nVersion), params));
+
+    BOOST_CHECK(CheckAuxProofOfWork(header, params));
 }
 
 BOOST_AUTO_TEST_CASE(auxpow_block800000_test) {
@@ -130,6 +133,8 @@ BOOST_AUTO_TEST_CASE(auxpow_block800000_test) {
     Consensus::Params params = CChainParams::Main({})->GetConsensus();
     BOOST_CHECK(auxpow.CheckAuxBlockHash(
         header.GetHash(), VersionChainId(header.nVersion), params));
+
+    BOOST_CHECK(CheckAuxProofOfWork(header, params));
 }
 
 BOOST_AUTO_TEST_CASE(auxpow_block3000000_test) {
@@ -155,6 +160,8 @@ BOOST_AUTO_TEST_CASE(auxpow_block3000000_test) {
     Consensus::Params params = CChainParams::Main({})->GetConsensus();
     BOOST_CHECK(auxpow.CheckAuxBlockHash(
         header.GetHash(), VersionChainId(header.nVersion), params));
+
+    BOOST_CHECK(CheckAuxProofOfWork(header, params));
 }
 
 BOOST_AUTO_TEST_CASE(auxpow_parse_coinbase_test) {
diff --git a/src/test/util/setup_common.cpp b/src/test/util/setup_common.cpp
index 7b876172d86a81be9505e1d74448938e5701b5ad..a3147b01d111c1d4f51761c5e46cb9f841424647 100644
--- a/src/test/util/setup_common.cpp
+++ b/src/test/util/setup_common.cpp
@@ -380,6 +380,81 @@ CBlock TestChain100Setup::CreateAndProcessBlock(
     return block;
 }
 
+CBlock TestChain100Setup::CreateAuxPowBlock(
+    const std::vector<CMutableTransaction> &txns, const CScript &scriptPubKey,
+    uint32_t parentChainId, uint32_t mergeMineNonce,
+    const std::vector<uint256> chainMerkleBranch,
+    const std::vector<uint256> coinbaseMerkleBranch, Chainstate &chainstate) {
+    const Config &config = GetConfig();
+    CBlock block = this->CreateBlock(txns, scriptPubKey, chainstate);
+
+    const Consensus::Params &params = config.GetChainParams().GetConsensus();
+    block.nNonce = 0;
+    // Ensure merge-mined block's nonce DOESNT mine the block
+    while (CheckProofOfWork(block.GetPowHash(), block.nBits, params)) {
+        ++block.nNonce;
+    }
+
+    block.nVersion = VersionWithAuxPow(block.nVersion, true);
+    block.auxpow.reset(new CAuxPow());
+
+    CAuxPow &auxpow = *block.auxpow;
+    auxpow.vChainMerkleBranch = chainMerkleBranch;
+    auxpow.vMerkleBranch = coinbaseMerkleBranch;
+    auxpow.nIndex = 0;
+    auxpow.nChainIndex =
+        CalcExpectedMerkleTreeIndex(mergeMineNonce, AUXPOW_CHAIN_ID, 0);
+
+    uint256 hashChainRoot = ComputeMerkleRootForBranch(
+        block.GetHash(), auxpow.vChainMerkleBranch, auxpow.nChainIndex);
+    std::reverse(hashChainRoot.begin(), hashChainRoot.end());
+
+    std::vector<uint8_t> coinbaseBytes = ToByteVector(MERGE_MINE_PREFIX);
+    coinbaseBytes.insert(coinbaseBytes.end(), hashChainRoot.begin(),
+                         hashChainRoot.end());
+    CDataStream ss{0, 0};
+    ss << uint32_t(1 << chainMerkleBranch.size());
+    ss << mergeMineNonce;
+    Span<const uint8_t> span = MakeUCharSpan(ss);
+    coinbaseBytes.insert(coinbaseBytes.end(), span.begin(), span.end());
+
+    CMutableTransaction tx;
+    tx.vin.resize(1);
+    tx.vin[0].scriptSig = CScript(coinbaseBytes.begin(), coinbaseBytes.end());
+
+    auxpow.coinbaseTx = MakeTransactionRef(tx);
+    auxpow.parentBlock.nVersion = MakeVersionWithChainId(parentChainId, 0);
+    auxpow.parentBlock.hashMerkleRoot = ComputeMerkleRootForBranch(
+        auxpow.coinbaseTx->GetHash(), auxpow.vMerkleBranch, 0);
+    // Ensure parent block's nonce DOES mine the block
+    while (!CheckProofOfWork(auxpow.parentBlock.GetPowHash(), block.nBits,
+                             params)) {
+        ++auxpow.parentBlock.nNonce;
+    }
+
+    return block;
+}
+
+CBlock TestChain100Setup::CreateAndProcessAuxPowBlock(
+    const std::vector<CMutableTransaction> &txns, const CScript &scriptPubKey,
+    uint32_t parentChainId, uint32_t mergeMineNonce,
+    const std::vector<uint256> chainMerkleBranch,
+    const std::vector<uint256> coinbaseMerkleBranch, Chainstate *chainstate) {
+    if (!chainstate) {
+        chainstate = &Assert(m_node.chainman)->ActiveChainstate();
+    }
+
+    const CBlock block = this->CreateAuxPowBlock(
+        txns, scriptPubKey, parentChainId, mergeMineNonce, chainMerkleBranch,
+        coinbaseMerkleBranch, *chainstate);
+    std::shared_ptr<const CBlock> shared_pblock =
+        std::make_shared<const CBlock>(block);
+    Assert(m_node.chainman)
+        ->ProcessNewBlock(shared_pblock, true, true, nullptr);
+
+    return block;
+}
+
 CMutableTransaction TestChain100Setup::CreateValidMempoolTransaction(
     CTransactionRef input_transaction, int input_vout, int input_height,
     CKey input_signing_key, CScript output_destination, Amount output_amount,
diff --git a/src/test/util/setup_common.h b/src/test/util/setup_common.h
index 6e2584f340acc51e75b718a07eb0931ad58422a9..1d58d7169356006d0c615789ec75dab38a4bfde3 100644
--- a/src/test/util/setup_common.h
+++ b/src/test/util/setup_common.h
@@ -139,6 +139,29 @@ struct TestChain100Setup : public TestingSetup {
     CBlock CreateBlock(const std::vector<CMutableTransaction> &txns,
                        const CScript &scriptPubKey, Chainstate &chainstate);
 
+    /**
+     * Like CreateAndProcessBlock, but uses a merge-mined block that uses PoW
+     * from another (here, nonexistent) chain.
+     */
+    CBlock
+    CreateAndProcessAuxPowBlock(const std::vector<CMutableTransaction> &txns,
+                                const CScript &scriptPubKey,
+                                uint32_t parentChainId, uint32_t mergeMineNonce,
+                                const std::vector<uint256> chainMerkleBranch,
+                                const std::vector<uint256> coinbaseMerkleBranch,
+                                Chainstate *chainstate = nullptr);
+
+    /**
+     * Like CreateBlock, but uses a merge-mined block that uses PoW from another
+     * (here, nonexistent) chain.
+     */
+    CBlock CreateAuxPowBlock(const std::vector<CMutableTransaction> &txns,
+                             const CScript &scriptPubKey,
+                             uint32_t parentChainId, uint32_t mergeMineNonce,
+                             const std::vector<uint256> chainMerkleBranch,
+                             const std::vector<uint256> coinbaseMerkleBranch,
+                             Chainstate &chainstate);
+
     //! Mine a series of new blocks on the active chain.
     void mineBlocks(int num_blocks);
 
diff --git a/test/functional/dogecoin_auxpow.py b/test/functional/dogecoin_auxpow.py
new file mode 100644
index 0000000000000000000000000000000000000000..0dc205deb2849ab11cdc787d1bfe92099f4e1f7f
--- /dev/null
+++ b/test/functional/dogecoin_auxpow.py
@@ -0,0 +1,226 @@
+# Copyright (c) 2024 The Bitcoin developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""
+Test whether Dogecoin auxpow blocks are accepted.
+"""
+
+import time
+
+from test_framework.address import P2SH_OP_TRUE
+from test_framework.blocktools import (
+    VERSION_CHAIN_ID_BITS,
+    create_block,
+    create_coinbase,
+)
+from test_framework.messages import (
+    MERGE_MINE_PREFIX,
+    VERSION_AUXPOW_BIT,
+    CAuxPow,
+    COutPoint,
+    CTxIn,
+    calc_expected_merkle_tree_index,
+    ser_uint256,
+    uint256_from_compact,
+)
+from test_framework.script import CScript
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import assert_equal
+
+
+class DogecoinAuxpowTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.setup_clean_chain = True
+        self.num_nodes = 1
+        self.extra_args = [[]]
+        self.rpc_timeout = 240
+
+    def run_test(self):
+        node = self.nodes[0]
+
+        mocktime = int(time.time())
+        node.setmocktime(mocktime)
+
+        height = 1
+        coinbase_tx = create_coinbase(height)
+        coinbase_tx.vout[0].scriptPubKey = P2SH_OP_TRUE
+        coinbase_tx.rehash()
+        block = create_block(
+            int(node.getbestblockhash(), 16), coinbase_tx, mocktime + 1100
+        )
+        block.nVersion = VERSION_CHAIN_ID_BITS | VERSION_AUXPOW_BIT | 5
+        block.rehash()
+        block_hash = bytes.fromhex(block.hash)
+
+        # nIndex = 1 not allowed
+        block.auxpow = CAuxPow()
+        block.auxpow.nIndex = 1
+        with node.assert_debug_log(["AuxPow nIndex must be 0"]):
+            assert_equal(node.submitblock(block.serialize().hex()), "high-hash")
+
+        # Parent chain ID 0x62 not allowed
+        block.auxpow = CAuxPow()
+        block.auxpow.parentBlock.nVersion = 0x620000
+        with node.assert_debug_log(["AuxPow parent has our chain ID"]):
+            assert_equal(node.submitblock(block.serialize().hex()), "high-hash")
+
+        # len(vChainMerkleBranch) >= 31 not allowed
+        block.auxpow = CAuxPow()
+        block.auxpow.vChainMerkleBranch = [0] * 31
+        with node.assert_debug_log(["AuxPow chain merkle branch too long"]):
+            assert_equal(node.submitblock(block.serialize().hex()), "high-hash")
+
+        # vMerkleBranch doesn't prove coinbase tx
+        block.auxpow = CAuxPow()
+        block.auxpow.parentBlock.hashMerkleRoot = 0x1234
+        with node.assert_debug_log(["AuxPow merkle root incorrect"]):
+            assert_equal(node.submitblock(block.serialize().hex()), "high-hash")
+
+        # Coinbase tx has no inputs
+        block.auxpow = CAuxPow()
+        block.auxpow.coinbaseTx.rehash()
+        block.auxpow.parentBlock.hashMerkleRoot = block.auxpow.coinbaseTx.sha256
+        with node.assert_debug_log(["AuxPow coinbase transaction missing input"]):
+            assert_equal(node.submitblock(block.serialize().hex()), "high-hash")
+
+        # Chain merkle root not found in coinbase scriptSig
+        block.auxpow = CAuxPow()
+        block.auxpow.coinbaseTx.vin = [CTxIn()]
+        block.auxpow.coinbaseTx.rehash()
+        block.auxpow.parentBlock.hashMerkleRoot = block.auxpow.coinbaseTx.sha256
+        with node.assert_debug_log(
+            ["AuxPow missing chain merkle root in parent coinbase"]
+        ):
+            assert_equal(node.submitblock(block.serialize().hex()), "high-hash")
+
+        # Duplicate merge mine prefix
+        block.auxpow = CAuxPow()
+        coinbase_script = CScript(MERGE_MINE_PREFIX + MERGE_MINE_PREFIX + block_hash)
+        block.auxpow.coinbaseTx.vin = [CTxIn(COutPoint(), coinbase_script)]
+        block.auxpow.coinbaseTx.rehash()
+        block.auxpow.parentBlock.hashMerkleRoot = block.auxpow.coinbaseTx.sha256
+        with node.assert_debug_log(["Multiple merged mining prefixes in coinbase"]):
+            assert_equal(node.submitblock(block.serialize().hex()), "high-hash")
+
+        # Padding bytes between merge mine prefix and chain merkle root
+        block.auxpow = CAuxPow()
+        coinbase_script = CScript(MERGE_MINE_PREFIX + b"hola" + block_hash)
+        block.auxpow.coinbaseTx.vin = [CTxIn(COutPoint(), coinbase_script)]
+        block.auxpow.coinbaseTx.rehash()
+        block.auxpow.parentBlock.hashMerkleRoot = block.auxpow.coinbaseTx.sha256
+        with node.assert_debug_log(
+            ["Merged mining prefix is not just before chain merkle root"]
+        ):
+            assert_equal(node.submitblock(block.serialize().hex()), "high-hash")
+
+        # Too many padding bytes before the chain merkle root
+        block.auxpow = CAuxPow()
+        coinbase_script = CScript(bytes(21) + block_hash)
+        block.auxpow.coinbaseTx.vin = [CTxIn(COutPoint(), coinbase_script)]
+        block.auxpow.coinbaseTx.rehash()
+        block.auxpow.parentBlock.hashMerkleRoot = block.auxpow.coinbaseTx.sha256
+        with node.assert_debug_log(
+            [
+                "AuxPow chain merkle root can have at most 20 preceding bytes of the parent coinbase"
+            ]
+        ):
+            assert_equal(node.submitblock(block.serialize().hex()), "high-hash")
+
+        # Missing nTreeSize and nMergeMineNonce (with prefix)
+        block.auxpow = CAuxPow()
+        coinbase_script = CScript(MERGE_MINE_PREFIX + block_hash)
+        block.auxpow.coinbaseTx.vin = [CTxIn(COutPoint(), coinbase_script)]
+        block.auxpow.coinbaseTx.rehash()
+        block.auxpow.parentBlock.hashMerkleRoot = block.auxpow.coinbaseTx.sha256
+        with node.assert_debug_log(
+            ["AuxPow missing chain merkle tree size and nonce in parent coinbase"]
+        ):
+            assert_equal(node.submitblock(block.serialize().hex()), "high-hash")
+
+        # Missing nTreeSize and nMergeMineNonce (without prefix)
+        block.auxpow = CAuxPow()
+        coinbase_script = CScript(block_hash)
+        block.auxpow.coinbaseTx.vin = [CTxIn(COutPoint(), coinbase_script)]
+        block.auxpow.coinbaseTx.rehash()
+        block.auxpow.parentBlock.hashMerkleRoot = block.auxpow.coinbaseTx.sha256
+        with node.assert_debug_log(
+            ["AuxPow missing chain merkle tree size and nonce in parent coinbase"]
+        ):
+            assert_equal(node.submitblock(block.serialize().hex()), "high-hash")
+
+        # Bad nTreeSize
+        block.auxpow = CAuxPow()
+        coinbase_script = CScript(
+            MERGE_MINE_PREFIX + block_hash + b"\xff\xff\xff\xff\xff\xff\xff\xff"
+        )
+        block.auxpow.coinbaseTx.vin = [CTxIn(COutPoint(), coinbase_script)]
+        block.auxpow.coinbaseTx.rehash()
+        block.auxpow.parentBlock.hashMerkleRoot = block.auxpow.coinbaseTx.sha256
+        with node.assert_debug_log(
+            ["AuxPow merkle branch size does not match parent coinbase"]
+        ):
+            assert_equal(node.submitblock(block.serialize().hex()), "high-hash")
+
+        block.auxpow = CAuxPow()
+        coinbase_script = CScript(
+            MERGE_MINE_PREFIX + block_hash + b"\x01\0\0\0\xff\xff\xff\xff"
+        )
+        block.auxpow.coinbaseTx.vin = [CTxIn(COutPoint(), coinbase_script)]
+        block.auxpow.coinbaseTx.rehash()
+        block.auxpow.parentBlock.hashMerkleRoot = block.auxpow.coinbaseTx.sha256
+
+        # Parent header has high hash
+        target = uint256_from_compact(block.nBits)
+        block.auxpow.parentBlock.rehashPow()
+        while block.auxpow.parentBlock.powHash < target:
+            block.auxpow.parentBlock.nNonce += 1  # un-solve block
+            block.auxpow.parentBlock.rehashPow()
+        with node.assert_debug_log(["Auxillary header proof of work failed"]):
+            assert_equal(node.submitblock(block.serialize().hex()), "high-hash")
+
+        # Success!
+        block.solve()
+        assert_equal(node.submitblock(block.serialize().hex()), None)
+        assert_equal(node.getbestblockhash(), block.hash)
+
+        # Add another block with each a merkle branch of 1 (instead of 0)
+        height = 2
+        coinbase_tx = create_coinbase(height)
+        coinbase_tx.vout[0].scriptPubKey = P2SH_OP_TRUE
+        coinbase_tx.rehash()
+        block2 = create_block(block.sha256, coinbase_tx, mocktime + 1101)
+        block2.nVersion = VERSION_CHAIN_ID_BITS | VERSION_AUXPOW_BIT | 0xFF
+        block2.rehash()
+        block_hash = bytes.fromhex(block.hash)
+
+        nMerkleNonce = 0xD4D3D2D1
+        nChainIndex = 1
+        while calc_expected_merkle_tree_index(nMerkleNonce, 0x62, 1) != nChainIndex:
+            nMerkleNonce += 1
+
+        block2.auxpow = CAuxPow()
+        chain_merkle_root = block2.get_merkle_root(
+            [ser_uint256(555), ser_uint256(block2.sha256)]
+        )
+        coinbase_script = CScript(
+            MERGE_MINE_PREFIX
+            + ser_uint256(chain_merkle_root)[::-1]
+            + b"\x02\0\0\0"
+            + nMerkleNonce.to_bytes(4, "little")
+        )
+        block2.auxpow.coinbaseTx.vin = [CTxIn(COutPoint(), coinbase_script)]
+        block2.auxpow.coinbaseTx.rehash()
+        block2.auxpow.vChainMerkleBranch = [555]
+        block2.auxpow.nChainIndex = 1
+        block2.auxpow.vMerkleBranch = [444]
+        block2.auxpow.parentBlock.hashMerkleRoot = block2.get_merkle_root(
+            [ser_uint256(block2.auxpow.coinbaseTx.sha256), ser_uint256(444)]
+        )
+        block2.solve()
+
+        assert_equal(node.submitblock(block2.serialize().hex()), None)
+        assert_equal(node.getbestblockhash(), block2.hash)
+
+
+if __name__ == "__main__":
+    DogecoinAuxpowTest().main()
diff --git a/test/functional/test_runner.py b/test/functional/test_runner.py
index 96d212dc412782211117e2881c5402816b980626..25f70b64b42017c0981f28dac11fbc1bf4949d44 100755
--- a/test/functional/test_runner.py
+++ b/test/functional/test_runner.py
@@ -867,7 +867,7 @@ def check_script_prefixes(all_scripts):
     LEEWAY = 0
 
     good_prefixes_re = re.compile(
-        "(abc_)?(example|feature|interface|mempool|mining|p2p|rpc|wallet|tool|chronik)_"
+        "(abc_)?(example|feature|interface|mempool|mining|p2p|rpc|wallet|tool|chronik|dogecoin)_"
     )
     bad_script_names = [
         script for script in all_scripts if good_prefixes_re.match(script) is None
